File: workers/python/rabbitmq_worker.py
import asyncio
import json
import os
import sys
import random
from datetime import datetime, timezone

sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'packages'))
from shared_python import ProcessedEvent

import aio_pika
from sqlmodel import Session, create_engine

DB_URL = os.getenv("DATABASE_URL", "postgresql://postgres:password@localhost:5432/mq_db")
engine = create_engine(DB_URL)

async def process_message(message: aio_pika.abc.AbstractIncomingMessage):
    group_id = "payment-group"
    
    # We use manual ack/nack, so we do not use the context manager message.process()
    try:
        event_dict = json.loads(message.body.decode('utf-8'))
        order_id = event_dict.get("order_id")
        published_at_str = event_dict.get("published_at")
        
        consumed_at = datetime.now(timezone.utc)
        published_at = datetime.fromisoformat(published_at_str.replace("Z", "+00:00"))
        latency_ms = int((consumed_at - published_at).total_seconds() * 1000)
        
        if os.getenv("SIMULATE_FAILURE", "false") == "true" and random.random() < 0.5:
            raise Exception("Simulated failure for DLQ routing")
            
        print(f" [x] {group_id} Received: {order_id} (Latency: {latency_ms}ms)")
        
        with Session(engine) as session:
            event_record = ProcessedEvent(
                event_id=order_id,
                group_id=group_id,
                mq_type="rabbitmq",
                data=json.dumps(event_dict),
                latency_ms=latency_ms
            )
            session.add(event_record)
            session.commit()
            
        await message.ack()
        
    except Exception as e:
        print(f" [!] {group_id} Error processing message: {e}")
        await message.reject(requeue=False)

async def consume_orders():
    amqp_url = os.getenv("RABBITMQ_URL", "amqp://guest:guest@localhost:5672/")
    group_id = "payment-group"
    
    connection = await aio_pika.connect_robust(amqp_url)
    
    async with connection:
        channel = await connection.channel()
        
        await channel.set_qos(prefetch_count=1)
        
        queue = await channel.declare_queue(
            "orders.queue",
            durable=True,
            arguments={
                "x-dead-letter-exchange": "dead-letter.exchange",
                "x-dead-letter-routing-key": "dlq"
            }
        )
        
        print(f"[*] Python Worker ({group_id}) started. Listening on {amqp_url}...")
        
        await queue.consume(process_message, no_ack=False)
        
        await asyncio.Future()

if __name__ == "__main__":
    asyncio.run(consume_orders())

File: workers/node/src/rabbitmq.worker.ts
import amqp from 'amqplib';
import { Client } from 'pg';
import type { ProcessedEvent } from '@mq/shared-node';

const DB_URL = process.env['DATABASE_URL'] ?? 'postgresql://postgres:password@localhost:5432/mq_db';
const RABBITMQ_URL = process.env['RABBITMQ_URL'] ?? 'amqp://guest:guest@localhost:5672/';
const GROUP_ID = 'inventory-group';

async function insertProcessedEvent(
    client: Client,
    event: Omit<ProcessedEvent, 'id' | 'processedAt'>
): Promise<void> {
    const query = `
        INSERT INTO processed_events (event_id, group_id, mq_type, data, latency_ms)
        VALUES ($1, $2, $3, $4, $5)
    `;
    await client.query(query, [
        event.eventId,
        event.groupId,
        event.mqType,
        JSON.stringify(event.data),
        event.latencyMs,
    ]);
}

async function runWorker() {
    // 1. DB 연결
    const dbClient = new Client({ connectionString: DB_URL });
    await dbClient.connect();
    console.log(`Connected to DB for ${GROUP_ID}`);

    // 2. RabbitMQ 연결
    const connection = await amqp.connect(RABBITMQ_URL);
    const channel = await connection.createChannel();
    
    // DLQ 설정 
    const dlqExchange = 'dead-letter.exchange';
    const dlqQueue = 'dead-letter.queue';
    
    await channel.assertExchange(dlqExchange, 'direct', { durable: true });
    await channel.assertQueue(dlqQueue, { durable: true });
    await channel.bindQueue(dlqQueue, dlqExchange, 'dlq');

    // Primary 설정
    const exchange = 'orders';
    const queue = 'orders.queue';
    
    await channel.assertExchange(exchange, 'direct', { durable: true });
    await channel.assertQueue(queue, {
        durable: true,
        arguments: {
            'x-dead-letter-exchange': dlqExchange,
            'x-dead-letter-routing-key': 'dlq'
        }
    });
    await channel.bindQueue(queue, exchange, 'order.created');

    // Prefetch 1
    await channel.prefetch(1);

    console.log(`[*] Node.js Worker (${GROUP_ID}) started. Listening on ${RABBITMQ_URL}...`);

    channel.consume(queue, async (msg) => {
        if (!msg) return;

        try {
            const eventData = JSON.parse(msg.content.toString()) as Record<string, unknown>;
            const orderId = (eventData['orderId'] ?? eventData['order_id']) as string;
            const publishedAtStr = (eventData['publishedAt'] ?? eventData['published_at']) as string;

            const consumedAt = new Date();
            const publishedAt = new Date(publishedAtStr);
            const latencyMs = consumedAt.getTime() - publishedAt.getTime();

            console.log(` [x] ${GROUP_ID} Received: ${orderId} (Latency: ${latencyMs}ms)`);

            await insertProcessedEvent(dbClient, {
                eventId: orderId,
                groupId: GROUP_ID,
                mqType: 'rabbitmq',
                data: eventData,
                latencyMs,
            });

            // 정상 처리 확인 -> ack
            channel.ack(msg);

        } catch (error) {
            console.error(` [!] ${GROUP_ID} Error processing message:`, error);
            // 에러 시 nack 수행. false = requeue 안 함 (DLQ로 보냄)
            channel.nack(msg, false, false);
        }
    }, {
        noAck: false // 수동 ack 사용
    });
}

runWorker().catch(err => {
    console.error('Error running Node.js Worker:', err);
    process.exit(1);
});
